### PR TITLE
Added JIT compilation of SortDescription

### PR DESCRIPTION
**Pull request by kitaisreal**
Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve performance or ORDER BY, MergeJoin, insertion into MergeTree using JIT compilation of sort columns comparator.

### CODE REVIEW

**Comment by kitaisreal** [comment]
Local tests.
```sql
CREATE TABLE test_table (id UInt64, value1 UInt64, value2 UInt64, value3 UInt64) ENGINE=MergeTree ORDER BY id;
INSERT INTO test_table SELECT number, rand64(), rand64(), rand64() FROM numbers(1200000)
```
```
./clickhouse-benchmark --query="SELECT * FROM test_table ORDER BY value1, value2 ASC, value3 DESC, value3 ASC FORMAT Null SETTINGS compile_sort_description=0;"
localhost:9000, queries 7, QPS: 7.210, RPS: 8651667.303, MiB/s: 264.028, result RPS: 0.000, result MiB/s: 0.000.
```
```
./clickhouse-benchmark --query="SELECT * FROM test_table ORDER BY value1, value2 ASC, value3 DESC, value3 ASC FORMAT Null SETTINGS compile_sort_description=1;"
localhost:9000, queries 20, QPS: 8.964, RPS: 10756737.139, MiB/s: 328.270, result RPS: 0.000, result MiB/s: 0.000.
```
For `Nullable` columns, or `Memory` engine performance improvements can be greater.

**Comment by kitaisreal** [comment]
@Mergifyio update

**Comment by kitaisreal** [comment]
@Mergifyio update

**Comment by kitaisreal** [comment]
@Mergifyio update

**Comment by kitaisreal** [comment]
@Mergifyio update

File: src/Core/Settings.h [hunk 1]
```diff
@@ -116,6 +116,8 @@ class IColumn;
     M(UInt64, min_count_to_compile_expression, 3, "The number of identical expressions before they are JIT-compiled", 0) \
     M(Bool, compile_aggregate_expressions, true, "Compile aggregate functions to native code.", 0) \
     M(UInt64, min_count_to_compile_aggregate_expression, 3, "The number of identical aggregate expressions before they are JIT-compiled", 0) \
+    M(Bool, compile_sort_description, true, "Compile sort description to native code.", 0) \
+    M(UInt64, min_count_to_compile_sort_description, 3, "The number of identical sort descriptions before they are JIT-compiled", 0) \
```
Comment (vdimir) [inline src/Core/Settings.h:120]: Haven't found where `min_count_to_compile_sort_description` is checked

File: src/Core/SortCursor.h [hunk 2]
```diff
@@ -164,17 +180,36 @@ struct SortCursor : SortCursorHelper<SortCursor>
...
+
+            auto sort_description_func_typed = reinterpret_cast<JITSortDescriptionFunc>(impl->desc.compiled_sort_description);
+            int res = sort_description_func_typed(lhs_pos, rhs_pos, impl->raw_sort_columns_data.data(), rhs.impl->raw_sort_columns_data.data()); /// NOLINT
+
+            if (res > 0)
+                return true;
+            if (res < 0)
+                return false;
+
+            return impl->order > rhs.impl->order;
```
Comment (vdimir) [inline src/Core/SortCursor.h:196]: just a matter of taste
```suggestion
            if (res != 0)
                return res;
            return impl->order > rhs.impl->order;
```

File: src/Core/SortCursor.h [hunk 3]
```diff
@@ -190,8 +225,26 @@ struct SimpleSortCursor : SortCursorHelper<SimpleSortCursor>
         const auto & desc = impl->desc[0];
         int direction = desc.direction;
         int nulls_direction = desc.nulls_direction;
-        int res = impl->sort_columns[0]->compareAt(lhs_pos, rhs_pos, *(rhs.impl->sort_columns[0]), nulls_direction);
-        return res != 0 && ((res > 0) == (direction > 0));
+
+        bool result = false;
+
+#if USE_EMBEDDED_COMPILER
+        if (impl->desc.compiled_sort_description && rhs.impl->desc.compiled_sort_description)
```
Comment (vdimir) [inline src/Core/SortCursor.h:232]: Should `impl->desc.compiled_sort_description` and `rhs.impl->desc.compiled_sort_description` be the same by construction? The `rhs.impl->desc.compiled_sort_description` is not used below, but there's no difference which instance to use, is it?

**Review by vdimir — APPROVED** [review]

File: src/Core/Settings.h (hunk 1 above)
Comment (kitaisreal) [reply src/Core/Settings.h:120]: Updated.

File: src/Core/SortCursor.h (hunk 2 above)
Comment (kitaisreal) [reply src/Core/SortCursor.h:196]: It is not the same. res can be 1, or -1, if we convert them to bool they both will be false. We can write res > 0, but current implementation is readable enought.

**Pull request merged**